### ISSUE TITLE
assert members lack JSDoc in IDE

### ISSUE THREAD

**Issue by turadg** [endojs/endo#1392]
# Problem to be solved

members of the `Assert` type don't show their jsdoc on the member properties.

For example, [FailTag has comprehensive documentation](https://github.com/Agoric/agoric-sdk/blob/6e3bc14aa0a77d8bdfad72ca8f0d8893c1f80b5f/packages/assert/src/types.js#L282-L338). But when hovering on the symbol it's all elided:
<img width="604" alt="Screenshot 2022-12-06 at 9 41 44 AM" src="https://user-images.githubusercontent.com/21505/205983558-83482ec5-5ae3-40b0-b7cc-06d32d8670c9.png">

# Design

One option that works is to keep the docstrings in the `d.ts` files too. The downside is having to keep them in sync. To solve that we could build the .js from .ts. If we need .js without .ts, then we'll have to move the `Assert` type out of .d.ts, but I'm not sure we can then express it in JSDoc since it will be nested.

**Comment by erights** [comment]
> that we could build the .js from .ts. 

That would either require us to
   * audit the ts-to-js compiler for preservation of our security properties, and re-audit when the ts compiler changes, or
   * manually review the js output of the compiler every time either the compiler or the ts sources change.

So let's continue looking for solutions elsewhere, like moving the explanations to .d.ts files.

**Comment by turadg** [comment]
> audit the ts-to-js compiler for preservation of our security properties, and re-audit when the ts compiler changes

Whatever is in that audit could be automated, no? Do we have a test suite of security properties that must be maintained by any syntax transformer? If not, that seems valuable to have.

For the actual JS code generation we could use a "strip only" transformer like esbuild. https://esbuild.github.io/content-types/#typescript-caveats

But I like what I think you're proposing: moving the JSDoc to TypeScript files.

**Comment by erights** [comment]
> Whatever is in that audit could be automated, no? Do we have a test suite of security properties that must be maintained by any syntax transformer? If not, that seems valuable to have.

Yes it could. On the one hand, that strikes me as a significant research project that would take much longer than the first audits, but would not need to be redone.

> For the actual JS code generation we could use a "strip only" transformer like esbuild. https://esbuild.github.io/content-types/#typescript-caveats

Yes, I would find a strip-only one acceptable. It would sidestep all my strong reasons for worry. But is the source to that truly TypeScript? Does the strip-only transform really preserve TypeScript's semantics? I haven't looked so I genuinely don't know. But this is interesting.

**Comment by erights** [comment]
> But I like what I think you're proposing: moving the JSDoc to TypeScript files.

Good. Let's do that for now while we continue exploring other options. Thanks!

**Comment by mhofman** [comment]
I am still against any solution that requires runtime code generation, and I also would prefer avoiding any solution that requires `.d.ts` file generation outside of a publish flow.

I do not fully understand where JSdoc fails here, and why it can't do something supported in `.d.ts`.
Or is the problem that we have a duplication of definitions between JSDoc and `.d.ts` that are not in sync ?

**Comment by SMotaal** [comment]
I've dug closer and as far as I can tell TypeScript's handling of JSDoc *"doc"* comment propagation is messy and has several issues that are bleak for lack of better words.

The one approach I considered was to use a `types.d.ts` instead of `types.js` and even then there is no clean way for the intuitive propagation to take place.

For instance, and please stop when you feel nauseated: *(see code below)*

1. I could get `const { Fail } = assert` to show the *"doc"* for `Fail` which come from `interface Assert {}` but this is the easy part.
2. I could get both `makeAssert` and `assert.makeAssert` to share the same *"doc"* which come the same `interface Assert {}` but only when I resort to ugly JavaScript code like `const { makeAssert } = /** @type {Assert} */ ({ makeAssert: localMakeAssert });` and this is the least ugly of the patterns I tried. 🙄
3. I could ***not*** get `const { Fail: Throws } = assert` to show the *"doc"* for `Throws` because somehow this functionality is labeled a "Suggestion" in microsoft/TypeScript#49798.

@mhofman At this point, it seems even serious refactoring and ugly patterns in the actual JavaScript modules will not deliver what is needed.

The key questions I've been considering:

1. Where is it more ideal for developers to maintain the *"doc"*?
   - The actual JavaScript modules? (not sufficient)
   - Using hand-coded ambient or imported `types.d.ts` in folders? (stable TypeScript)
   - ~~Using hand-coded ambient or imported `types.js` in folders?~~ (buggy JSDoc)
   - ~~Using a generated `types.d.ts` in folders?~~ (too many moving parts)
2. What is the most suitable way to propagate the *"doc"*?
   - Language server plugin?
   - Publish flow?

I think for the top question it is likely some conventional patterns may be needed to determine which *"doc"* comments end up in JavaScript modules and which ones will need to be propagated from elsewhere. It seems probable at this point that a generated `types.d.ts` at the package scope level may be the way to go.

The second question is ***not*** something I am ready to weigh in at this point, but would appreciate ideas I could investigate further on my end.

@turadg Did you have ideas I could explore on my end?

---

<details><summary>Code</summary>

A minimal example of the ambient `types.d.ts` in folder:

```ts
/// @filename: packages/ses/src/error/types.d.ts

/** doc for `Assert` interface itself */
interface Assert {
  /** doc for `(@type {Assert} ({ ... })).Fail` */
  Fail: FailTag;

  /** doc for `(@type {Assert} ({ ... })).makeAssert` */
  makeAssert: MakeAssert
}
```

A minimal example of the refactored exporter module:

```js
/// @filename: packages/ses/src/error/assert.js
/// <reference path="./types.d.ts" />

import { makeAssert as internalMakeAssert } from './internal/makeAssert.js';

const { makeAssert } = /** @type {Pick<Assert, 'makeAssert>} */ ({
  makeAssert: internalMakeAssert
}));

export {makeAssert};

/** doc for exported `assert` **/
const assert = makeAssert();

export { assert };
```

A pseudo test just to give an idea of what ends up happening:

```ts
/// @filename: packages/ses/test/types/src/error/test-assert-types.ts

import { assert, makeAssert } from "../../../../src/error/assert.js";

// `assert` doc is: "doc for exported `assert`"
// `makeAssert` doc is: "doc for `(@type {Assert} ({ ... })).makeAssert`"

{
  const { makeAssert } = assert;

  // `assert.makeAssert` doc is: "doc for `(@type {Assert} ({ ... })).makeAssert`"
  // `makeAssert` doc is: "doc for `(@type {Assert} ({ ... })).makeAssert`"
}

{
  const { Fail, Fail: Throws } = assert;

  assert.Fail`…`;
  Fail`…`;
  Throws`…`;

  // `assert.Fail` doc is: "doc for `(@type {Assert} ({ ... })).Fail`"
  // `Fail` doc is: "doc for `(@type {Assert} ({ ... })).Fail`"
  // `Throws` doc is: missing!
}
```

</details>

**Comment by turadg** [comment]
> @turadg Did you have ideas I could explore on my end?

Nope. There's no silver bullet so any improvement will be a tradeoff of competing concerns. I expect we'll improve incrementally in small steps. If a clear path forward for the codebase emerges, I'll try to document here.

**Comment by SMotaal** [comment]
I went ahead created an experimental branch to try to explore this further.

I decided to reduce the clutter and start fresh in [`endo/retype/`](https://github.com/SMotaal/endo/tree/smotaal/retyped-ses-experiment/retyped) where I worked on a minimal example of how to ensure the JSDoc doc would propagate.

Being further out and going with what is reflected in threads and code, I hope my efforts spark some discussions, which would allow me to better understand the tradeoffs and concerns.

Based on my understanding so far, the pattern I am using tries to ensures that there is a single source of truth and that it is JavaScript code itself.

This pattern still requires a package-level [`types.d.ts`](https://github.com/SMotaal/endo/blob/smotaal/retyped-ses-experiment/retyped/packages/retyped-ses/types.d.ts) file for consumers, which is generated with `tsc` and can theoretically be a `prepare` or similar package script.

The funny thing about TypeScript is that there are not so subtle differences in the types it will generate from subtle differences in the JavaScript code which ultimately would not have any observable runtime differences.

That said, sometimes it takes more JavaScript to make sure it is happy, and the one example I included to make this point is the non-constructable [`DetailsToken`](https://github.com/SMotaal/endo/blob/261bf0eb7411f94070c554f526e685a3b17e4751/retyped/packages/retyped-ses/src/error/internal/details.js#L7-L24) pre-es2015-style class.

@turadg Can you check my branch out on your end and let me know what you think? Please see [`endo/retyped/README.md`](https://github.com/SMotaal/endo/blob/smotaal/retyped-ses-experiment/retyped/README.md) for additional details.

**Comment by turadg** [comment]
> @turadg Can you check out what I have so far in [`endo/retyped`](https://github.com/SMotaal/endo/tree/smotaal/retyped-ses-experiment/retyped) on your end and let me know what you think?

It's a little hard for me to follow. I understand why you'd want to have a clean package instead of modifying in place, but it does make it harder to see how it would map back to solving the problem in the issue description.

What would be ideal to me is a PR that changes as little as possible while ensuring that Assert members have jsdoc strings, including a screenshot demonstrating that for `Fail`.

**Comment by kriskowal** [comment]
@turadg Has this issue been addressed to your satisfaction? Are the next steps for `@endo/errors` tracked elsewhere? I recall a naming concern with @erights and I’m not sure we settled that. We also presumably should be tracking adoption in Agoric SDK, since I see that hasn’t begun.

**Comment by turadg** [comment]
> Are the next steps for @endo/errors tracked elsewhere?

I'm not sure. https://github.com/endojs/endo/pull/1717#discussion_r1297908833 had some suggestions. I think https://github.com/endojs/endo/pull/1729/files implemented them.

> Has this issue been addressed to your satisfaction?

I'm not aware of it being addressed at all. The `Assert` type still lacks documentation: https://github.com/endojs/endo/blob/d8a7dcc766b1d7ae7dd676943e8e97a7cce5e56d/packages/ses/types.d.ts#L211-L222

The `@endo/errors` work didn't solve that.

**Comment by erights** [comment]
>  I recall a naming concern with @erights and I’m not sure we settled that.

Anyone remember what that was?
